File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
def solution(land):
    
    # 가로열의 최소열 최대열만 구하면 세로로 더하지 않아도 된다.
    n, m = len(land), len(land[0]);
    visited = [[False] * m for _ in range(n)]
    direction = [
        [0,1],
        [1,0],
        [0,-1],
        [-1,0]
    ];
    
    def BFS(sx, sy) :
        queue = [(sx, sy)]
        visited[sx][sy] = True;
        values = [0,500,0];
        count = 1;
        
        while len(queue) > 0:
            cx, cy = queue.pop();
            if(cy<values[1]): values[1] = cy;
            if(cy>values[2]): values[2] = cy;
            
            for [dx, dy] in direction:
                nx, ny = dx+cx, dy+cy;
                if(0<=nx<n and 0<=ny<m and not visited[nx][ny] and land[nx][ny] == 1):
                    visited[nx][ny] = True;
                    count+=1;
                    queue.append((nx, ny))
        values[0] = count;
        return values;
    
    answer = [0]*m;
    for i in range(n):
        for j in range(m):
            if not visited[i][j] and land[i][j] == 1:
                value, start, end = BFS(i, j);
                for idx in range(start, end+1):
                    answer[idx] += value;
                    
    return max(answer);
        

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js
const solution = (land) => {
    const n = land.length;
    const m = land[0].length;
    
    visited = Array.from({length: n}, () => Array(m).fill(false));
    const oilLand = Array.from({length: n}, () => Array(m).fill(0));;
    const groupMap = new Map();
    const direction = [
        [1,0],
        [0,1],
        [-1,0],
        [0,-1]
    ]
    let groupId = 1;
    
    const BFS = (cx, cy) => {
        if (visited[cx][cy] === true || land[cx][cy] === 0) return 0;    
        let [count,front] = [1,0];   
        const queue = [[cx, cy]];  
        const group = [[cx, cy]];
        visited[cx][cy] = true;
        
        while (front < queue.length){
            const [x, y] = queue[front++];
        
            for (const [dx, dy] of direction){
                const [nx, ny] = [dx+x, dy+y];
                
                if(0<=nx && nx<n && 0<= ny && ny<m && land[nx][ny] === 1 && !visited[nx][ny]){
                    visited[nx][ny] = true;
                    group.push([nx, ny]);
                    queue.push([nx, ny]);
                    count += 1;
                }
            }   
        }
        for (const [px, py] of group){
            if (px >= 0 && px < n && py >= 0 && py < m) {
                oilLand[px][py] = groupId;
            }
        }
        groupMap.set(groupId, count);
        groupId += 1;
        return count;
    }
    
    for (let i=0; i<n; i++){
        for (let j=0; j<m; j++){
            if (!visited[i][j] && land[i][j] === 1) {
                BFS(i, j);
            }
        }
    }
    
    //console.table(oilLand);
    //console.log(groupMap);
    
    const colSums = Array(m).fill(0);
    for (let j=0; j<m; j++){
        const seen = new Set();
        for (let i=0; i<n; i++){
            const groupId = oilLand[i][j]; 
            if (groupId === 0 || seen.has(groupId)) continue;
            colSums[j] += groupMap.get(groupId);
            seen.add(groupId);
        }
    }
    
    // console.table(oilLand);
    const maxSum = Math.max(...colSums);
    return maxSum;
}